doc/architectural/central-data-structures.md

```diff
@@ -0,0 +1,100 @@
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) used
+for the GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a pair, consisting of:
+
+* A list of GOTO functions,
+* A symbol table containing symbol references for the symbols contained in the GOTO functions.
+
+In pseudocode, the type looks this:
+
+```js
+type goto_modelt {
+  type goto_functionst = list<goto_functiont>
+  type symbol_tablet = map<identifier, symbolt>
+}
+```
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also defined as a pair. It's designed to represent a function
+at the IR level, and effectively it's the following data type (in pseudocode):
+
+```js
+type goto_functiont {
+    goto_programt body
+    list<identifiers> parameters
+}
+```
+
+The `goto_programt` denoting the `body` of the function will be the subject of
+a more elaborate explanation in the next section.
+
+The `parameters` subcomponent is a list of identifiers that are to be looked-up
+in the symbol-table for their values.
+
+## goto_programt
+
+A `goto_programt` is a list of GOTO-IR instructions. In pseudocode, it would
+look like `type goto_programt = list<goto_instructiont>`.
+
+An instruction (`goto_instructiont`) is a triple (an element with three subcomponents),
+describing a GOTO level instruction with the following 3 component subtypes,
+again in pseudocode:
+
+```js
+type goto_instructiont {
+    instruction_type instr_type
+    instruction      statement
+    guard            boolean_expr
+}
+```
+
+The pseudocode subtypes above require some more elaboration, so we will provide some extra
+detail to illuminate some details at the code level:
+
+* The `instruction_type` above corresponds to the `goto_program_instruction_typet` type
+  listed in [`goto_program.h`](../../src/goto-programs/goto_program.h)
+  * Some illustrative instruction types are `assign`, `function_call`, `return`, etc.
+* The `instruction` is a statement represented by a `goto_instruction_codet`.
+  * A `goto_instruction_codet` is an alias of `codet` (a data structure broadly representing
+    a statement inside CBMC) that contains the actual code to be executed.
+  * You can distinguish different statements by inspecting the `irep` element `ID_statement`.
+* The `guard` is an `exprt` (a data structure broadly representing an expression inside CBMC)
+  that is expected to have type `bool`.
+  * This is optional - not every instruction is expected to have a `guard` associated with it.
+
+## source_locationt
+
+Another important data structure that needs to be discussed at this point is
+`source_locationt`.
+
+This is an `irept`. `irep`s are the central data structure that model most entities inside
```

Comment:
Is the fact that `source_locationt` inherits from `irept` relevant to this part of the documentation? I would consider this to be a low-level implementation detail and omit it from the high level overview.

Reply:
Agreed, but it might be important to cover `irept` by itself?

Reply:
I have separated the explanation of `irept` itself and added further details about it.